// zarm/lib/mask/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

var maskStyles = _reactNative.StyleSheet.create(_index.default);

var Mask =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Mask, _PureComponent);

  function Mask() {
    (0, _classCallCheck2.default)(this, Mask);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Mask).apply(this, arguments));
  }

  (0, _createClass2.default)(Mask, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          visible = _this$props.visible,
          styles = _this$props.styles,
          type = _this$props.type,
          style = _this$props.style,
          onClick = _this$props.onClick;
      var popupCls = [styles.wrapperStyle, styles["".concat(type, "Wrapper")], style];
      return visible && _react.default.createElement(_reactNative.TouchableWithoutFeedback, {
        onPress: onClick
      }, _react.default.createElement(_reactNative.View, {
        style: [popupCls]
      }));
    }
  }]);
  return Mask;
}(_react.PureComponent);

exports.default = Mask;
Mask.defaultProps = {
  visible: false,
  type: 'normal',
  styles: maskStyles
};
