// zarm/lib/alert/Alert.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _modal = _interopRequireDefault(require("../modal"));

var _zh_CN = _interopRequireDefault(require("./locale/zh_CN"));

var Alert =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Alert, _PureComponent);

  function Alert() {
    (0, _classCallCheck2.default)(this, Alert);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Alert).apply(this, arguments));
  }

  (0, _createClass2.default)(Alert, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          content = _this$props.content,
          cancelText = _this$props.cancelText,
          onCancel = _this$props.onCancel,
          locale = _this$props.locale,
          others = (0, _objectWithoutProperties2.default)(_this$props, ["prefixCls", "className", "content", "cancelText", "onCancel", "locale"]);
      var cls = (0, _classnames.default)(prefixCls, className);
      return _react.default.createElement(_modal.default, (0, _extends2.default)({
        className: cls
      }, others, {
        footer: _react.default.createElement("button", {
          type: "button",
          className: "".concat(prefixCls, "__button"),
          onClick: onCancel
        }, cancelText || locale.cancelText)
      }), content);
    }
  }]);
  return Alert;
}(_react.PureComponent);

exports.default = Alert;
Alert.defaultProps = {
  prefixCls: 'za-alert',
  animationType: 'zoom',
  locale: _zh_CN.default
};
