// zarm/lib/file-picker/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _handleFileInfo = _interopRequireDefault(require("./utils/handleFileInfo"));

/**
 * 设置选择前的方法，获取选择文件的相关信息，需要时可以对图片进行压缩、改变图片尺寸。
 *
 * multiple: 默认为 false，设置为 true 之后一次可以选择多张，onChange 事件调用之后返回一个数组，
 *           不设置或者设置为 false，onChange 事件调用之后返回一个对象。
 * disabled: 传递之后不可以点击上传，整个选择组件会设置为半透明状态，透明度为 0.5。
 * quality: 没有默认值，不设置不会进行压缩。
 * accept: 设置选择的文件类型，默认为所有类型，只有文件类型为图片（image/*）的时候会有本地预览图。
 * onChange: () => { file, fileType, fileSize, fileName, thumbnail }。
 * onBeforeSelect: () => boolean，返回 false 的时候阻止后续的选择事件。
 */
var FilePicker =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(FilePicker, _PureComponent);

  function FilePicker() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, FilePicker);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(FilePicker)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.file = null;

    _this.handleDefaultInput = function (e) {
      // 防止选择同一张图片两次造成 onChange 事件不触发
      e.currentTarget.value = '';
      var _this$props = _this.props,
          onBeforeSelect = _this$props.onBeforeSelect,
          disabled = _this$props.disabled;

      if (typeof onBeforeSelect !== 'function') {
        return;
      } // 阻止 input onChange 默认事件


      if (onBeforeSelect() === false || disabled) {
        e.preventDefault();
      }
    };

    _this.handleClick = function () {
      _this.file.click();
    };

    _this.handleChange = function (e) {
      var _this$props2 = _this.props,
          onChange = _this$props2.onChange,
          quality = _this$props2.quality,
          multiple = _this$props2.multiple;
      var files = [].slice.call(e.target.files);
      var fileList = [];

      var getFileInfo = function getFileInfo(data) {
        if (multiple) {
          fileList.push(data);

          if (files.length === fileList.length && typeof onChange === 'function') {
            onChange(fileList);
          }
        } else {
          typeof onChange === 'function' && onChange(data);
        }
      };

      if (files) {
        files.map(function (file) {
          return (0, _handleFileInfo.default)({
            file: file,
            quality: quality
          }, getFileInfo);
        });
      }
    };

    return _this;
  }

  (0, _createClass2.default)(FilePicker, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props3 = this.props,
          prefixCls = _this$props3.prefixCls,
          className = _this$props3.className,
          multiple = _this$props3.multiple,
          accept = _this$props3.accept,
          capture = _this$props3.capture,
          disabled = _this$props3.disabled,
          children = _this$props3.children;
      var cls = (0, _classnames.default)(prefixCls, className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "--disabled"), disabled));
      var content = (0, _react.cloneElement)(children, {
        onClick: this.handleClick,
        className: 'needsclick' // 修复加载fastClick库后引起的合成事件问题

      });
      return _react.default.createElement("div", {
        className: cls
      }, _react.default.createElement("input", {
        className: "".concat(prefixCls, "__input"),
        type: "file",
        ref: function ref(ele) {
          _this2.file = ele;
        },
        accept: accept,
        multiple: multiple,
        capture: capture,
        onClick: this.handleDefaultInput,
        onChange: this.handleChange
      }), content);
    }
  }]);
  return FilePicker;
}(_react.PureComponent);

exports.default = FilePicker;
FilePicker.defaultProps = {
  prefixCls: 'za-file-picker',
  disabled: false,
  multiple: false,
  onBeforeSelect: function onBeforeSelect() {
    return true;
  }
};
