// zarm/es/keyboard/Keyboard.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React, { PureComponent } from 'react';
import classnames from 'classnames';
import Icon from '../icon';
var KEYS = {
  number: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', 'close'],
  price: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', 'close'],
  idcard: ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'X', '0', 'close']
};

var Keyboard =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Keyboard, _PureComponent);

  function Keyboard() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Keyboard);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Keyboard)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.longPressTimer = void 0;

    _this.onLongPressIn = function (key) {
      _this.onKeyClick(key);

      _this.longPressTimer = setTimeout(function () {
        _this.longPressTimer = setInterval(function () {
          _this.onKeyClick(key);
        }, 100);
      }, 800);
    };

    _this.onLongPressOut = function (e) {
      e.preventDefault();
      clearInterval(_this.longPressTimer);
    };

    _this.onKeyClick = function (key) {
      if (key.length === 0) {
        return;
      }

      var onKeyClick = _this.props.onKeyClick;

      if (typeof onKeyClick === 'function') {
        onKeyClick(key);
      }
    };

    _this.getKeys = function () {
      var type = _this.props.type;
      return type ? KEYS[type] : KEYS.number;
    };

    _this.renderKey = function (text, index) {
      var prefixCls = _this.props.prefixCls;
      var keyCls = classnames("".concat(prefixCls, "__item"), _defineProperty({}, "".concat(prefixCls, "__item--disabled"), text.length === 0));
      return React.createElement("div", {
        className: keyCls,
        key: +index,
        onClick: function onClick() {
          return _this.onKeyClick(text);
        }
      }, text === 'close' ? React.createElement(Icon, {
        type: "keyboard",
        size: "lg"
      }) : text);
    };

    return _this;
  }

  _createClass(Keyboard, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          locale = _this$props.locale;
      return React.createElement("div", {
        className: prefixCls
      }, React.createElement("div", {
        className: "".concat(prefixCls, "__keys")
      }, this.getKeys().map(this.renderKey)), React.createElement("div", {
        className: "".concat(prefixCls, "__handle")
      }, React.createElement("div", {
        className: "".concat(prefixCls, "__item"),
        onTouchStart: function onTouchStart() {
          return _this2.onLongPressIn('delete');
        },
        onTouchEnd: this.onLongPressOut,
        onTouchCancel: this.onLongPressOut,
        onMouseDown: function onMouseDown() {
          return _this2.onLongPressIn('delete');
        },
        onMouseUp: this.onLongPressOut
      }, React.createElement(Icon, {
        type: "deletekey",
        size: "lg"
      })), React.createElement("div", {
        className: "".concat(prefixCls, "__item ").concat(prefixCls, "__item--ok"),
        onClick: function onClick() {
          return _this2.onKeyClick('ok');
        }
      }, locale.okText)));
    }
  }]);

  return Keyboard;
}(PureComponent);

Keyboard.defaultProps = {
  prefixCls: 'za-keyboard',
  type: 'number'
};
export { Keyboard as default };
