// zarm/es/panel/index.native.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React, { PureComponent } from 'react';
import { StyleSheet, View } from 'react-native';
import panelStyle from './style/index.native';
import { RenderWithText } from '../utils/renderWithText.native';
var panelStyles = StyleSheet.create(panelStyle);

var Panel =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Panel, _PureComponent);

  function Panel() {
    _classCallCheck(this, Panel);

    return _possibleConstructorReturn(this, _getPrototypeOf(Panel).apply(this, arguments));
  }

  _createClass(Panel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          more = _this$props.more,
          style = _this$props.style,
          styles = _this$props.styles,
          children = _this$props.children;
      var wrapperStyle = [styles.container, style];
      var headerStyle = [styles.panelHeader];
      var bodyStyle = [styles.panelBody];
      return React.createElement(View, {
        style: wrapperStyle
      }, React.createElement(View, {
        style: headerStyle
      }, React.createElement(RenderWithText, {
        component: title,
        viewStyle: panelStyles.panelHeaderTitle,
        textStyle: panelStyles.panelHeaderTitleText
      }), React.createElement(RenderWithText, {
        component: more,
        viewStyle: panelStyles.panelHeaderMore,
        textStyle: panelStyles.panelHeaderMoreText
      })), React.createElement(View, {
        style: bodyStyle
      }, children));
    }
  }]);

  return Panel;
}(PureComponent);

Panel.defaultProps = {
  styles: panelStyles
};
export { Panel as default };
