// zarm/lib/badge/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

var badgeStyles = _reactNative.StyleSheet.create(_index.default);

var Badge =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Badge, _PureComponent);

  function Badge() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Badge);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Badge)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.state = {
      dotWidth: 0
    };

    _this.layout = function (e) {
      var dotWidth = _this.props.shape === 'dot' || _this.props.shape === undefined ? -(parseInt(e.layout.width, 10) - 4) : -(parseInt(e.layout.width, 10) - 8);

      _this.setState({
        dotWidth: dotWidth
      });
    };

    return _this;
  }

  (0, _createClass2.default)(Badge, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          theme = _this$props.theme,
          shape = _this$props.shape,
          text = _this$props.text,
          style = _this$props.style,
          children = _this$props.children,
          styles = _this$props.styles;
      var dotWidth = this.state.dotWidth;
      var bagdeWrapper = [styles.textStyle, style];
      var badgeText = [styles.badgeText, styles["".concat(shape, "Text")], shape === undefined && styles.dotText];
      var iconStyle = [styles["".concat(shape, "Badge")], styles["".concat(theme, "Bagde")], !!children && [badgeStyles.sup], shape === undefined && styles.dotBadge];
      return _react.default.createElement(_reactNative.View, {
        style: bagdeWrapper
      }, children, _react.default.createElement(_reactNative.View, {
        onLayout: function onLayout(_ref) {
          var e = _ref.nativeEvent;
          return _this2.layout(e);
        },
        style: [iconStyle, {
          right: children ? dotWidth : 0
        }]
      }, _react.default.createElement(_reactNative.Text, {
        style: badgeText
      }, text)));
    }
  }]);
  return Badge;
}(_react.PureComponent);

exports.default = Badge;
Badge.defaultProps = {
  theme: 'danger',
  styles: badgeStyles
};
