// zarm/es/switch/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React, { PureComponent } from 'react';
import classnames from 'classnames';

var getChecked = function getChecked(props, defaultChecked) {
  if (typeof props.checked !== 'undefined') {
    return props.checked;
  }

  if (typeof props.defaultChecked !== 'undefined') {
    return props.defaultChecked;
  }

  return defaultChecked;
};

var Switch =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Switch, _PureComponent);

  function Switch() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Switch);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Switch)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.state = {
      checked: getChecked(_this.props, false)
    };

    _this.onValueChange = function () {
      var _this$props = _this.props,
          disabled = _this$props.disabled,
          onChange = _this$props.onChange;
      var checked = _this.state.checked;

      if (disabled) {
        return;
      }

      var newChecked = !checked;

      if (!('checked' in _this.props)) {
        _this.setState({
          checked: newChecked
        });
      }

      if (typeof onChange === 'function') {
        onChange(newChecked);
      }
    };

    return _this;
  }

  _createClass(Switch, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          prefixCls = _this$props2.prefixCls,
          className = _this$props2.className,
          disabled = _this$props2.disabled,
          style = _this$props2.style;
      var checked = this.state.checked;
      var cls = classnames(prefixCls, className, _defineProperty({}, "".concat(prefixCls, "--disabled"), disabled));
      return React.createElement("span", {
        className: cls,
        style: style
      }, React.createElement("input", {
        type: "checkbox",
        className: "".concat(prefixCls, "__input"),
        disabled: disabled,
        checked: checked,
        onChange: this.onValueChange,
        value: checked ? 'on' : 'off'
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps) {
      if (typeof nextProps.checked !== 'undefined') {
        return {
          checked: nextProps.checked
        };
      }

      return null;
    }
  }]);

  return Switch;
}(PureComponent);

Switch.displayName = 'Switch';
Switch.defaultProps = {
  prefixCls: 'za-switch',
  disabled: false
};
export { Switch as default };
