// zarm/es/alert/Alert.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React, { PureComponent } from 'react';
import classnames from 'classnames';
import Modal from '../modal';
import alertLocale from './locale/zh_CN'; // import { getRunTimeLocale } from '../locale-provider/LocaleProvider';

var Alert =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Alert, _PureComponent);

  function Alert() {
    _classCallCheck(this, Alert);

    return _possibleConstructorReturn(this, _getPrototypeOf(Alert).apply(this, arguments));
  }

  _createClass(Alert, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          content = _this$props.content,
          cancelText = _this$props.cancelText,
          onCancel = _this$props.onCancel,
          locale = _this$props.locale,
          others = _objectWithoutProperties(_this$props, ["prefixCls", "className", "content", "cancelText", "onCancel", "locale"]);

      var cls = classnames(prefixCls, className);
      return React.createElement(Modal, _extends({
        className: cls
      }, others, {
        footer: React.createElement("button", {
          type: "button",
          className: "".concat(prefixCls, "__button"),
          onClick: onCancel
        }, cancelText || locale.cancelText)
      }), content);
    }
  }]);

  return Alert;
}(PureComponent);

Alert.defaultProps = {
  prefixCls: 'za-alert',
  animationType: 'zoom',
  locale: alertLocale
};
export { Alert as default };
