// zarm/lib/message/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _index = _interopRequireDefault(require("./style/index.native"));

var _renderWithText = require("../utils/renderWithText.native");

var messageStyles = _reactNative.StyleSheet.create(_index.default);

var Message =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Message, _PureComponent);

  function Message(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Message);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Message).call(this, props));

    _this.onPressClose = function () {
      _this.setState({
        visible: false
      });
    };

    _this.state = {
      visible: true
    };
    return _this;
  }

  (0, _createClass2.default)(Message, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          theme = _this$props.theme,
          icon = _this$props.icon,
          hasArrow = _this$props.hasArrow,
          closable = _this$props.closable,
          styles = _this$props.styles,
          style = _this$props.style,
          onClick = _this$props.onClick,
          children = _this$props.children,
          size = _this$props.size;
      var visible = this.state.visible;
      var wrapperStyle = [styles.wrapper, styles["".concat(theme, "Wrapper")], styles["".concat(size, "Wrapper")], style];
      var textStyle = [styles["".concat(theme, "TextStyle")], styles["".concat(size, "TextStyle")]];
      var arrowStyle = [styles.arrowStyle, styles["".concat(theme, "ArrowStyle")], styles["".concat(size, "ArrowStyle")]];
      var closeIconStyle = [styles.closeIconStyle, styles["".concat(theme, "CloseIconStyle")], styles["".concat(size, "CloseIconStyle")]];

      var closeRender = closable && _react.default.createElement(_reactNative.TouchableWithoutFeedback, {
        onPress: this.onPressClose
      }, _react.default.createElement(_reactNative.View, {
        style: styles.closeIconWrapperStyle
      }, _react.default.createElement(_reactNative.View, {
        style: [closeIconStyle, styles.closeIconLeft]
      }), _react.default.createElement(_reactNative.View, {
        style: [closeIconStyle, styles.closeIconRight]
      })));

      var arrowRender = hasArrow && _react.default.createElement(_reactNative.View, {
        style: styles.arrowWrapperStyle
      }, _react.default.createElement(_reactNative.View, {
        style: arrowStyle
      }));

      var footerRender = (closable || hasArrow) && _react.default.createElement(_reactNative.View, {
        style: styles.footerStyle
      }, closeRender, arrowRender);

      var messageRender = _react.default.createElement(_reactNative.View, {
        style: wrapperStyle
      }, icon && _react.default.createElement(_reactNative.View, {
        style: styles.headerStyle
      }), _react.default.createElement(_renderWithText.RenderWithText, {
        viewStyle: styles.bodyStyle,
        textStyle: textStyle,
        component: children
      }), footerRender);

      var wrapperProps = {
        onPress: onClick
      };
      return visible && (onClick ? _react.default.createElement(_reactNative.TouchableWithoutFeedback, wrapperProps, messageRender) : messageRender);
    }
  }]);
  return Message;
}(_react.PureComponent);

exports.default = Message;
Message.defaultProps = {
  theme: 'primary',
  size: 'md',
  hasArrow: false,
  closable: false,
  styles: messageStyles
};
