// zarm/es/modal/ModalHeader.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React, { PureComponent } from 'react';
import classnames from 'classnames';
import Icon from '../icon';

var ModalHeader =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(ModalHeader, _PureComponent);

  function ModalHeader() {
    _classCallCheck(this, ModalHeader);

    return _possibleConstructorReturn(this, _getPrototypeOf(ModalHeader).apply(this, arguments));
  }

  _createClass(ModalHeader, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          title = _this$props.title,
          closable = _this$props.closable,
          onCancel = _this$props.onCancel,
          others = _objectWithoutProperties(_this$props, ["prefixCls", "className", "title", "closable", "onCancel"]);

      var cls = classnames("".concat(prefixCls, "__header"), className);
      var btnClose = closable && React.createElement(Icon, {
        type: "wrong",
        size: "sm",
        className: "".concat(prefixCls, "__header__close"),
        onClick: onCancel
      });
      return React.createElement("div", _extends({
        className: cls
      }, others), React.createElement("div", {
        className: "".concat(prefixCls, "__header__title")
      }, title), btnClose);
    }
  }]);

  return ModalHeader;
}(PureComponent);

ModalHeader.defaultProps = {
  prefixCls: 'za-modal'
};
export { ModalHeader as default };
