// zarm/lib/locale-receiver/LocaleReceiver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _LocaleProvider = require("../locale-provider/LocaleProvider");

var _zh_CN = _interopRequireDefault(require("../locale-provider/locale/zh_CN"));

var LocaleReceiverWrapper = function LocaleReceiverWrapper(name, defaultLocale) {
  return function InnerWrapper(WrappedComponent) {
    var LocaleReceiver = function LocaleReceiver(props) {
      return _react.default.createElement(_LocaleProvider.LocaleContext.Consumer, null, function (locale) {
        var globalLocale = locale.locale ? locale : defaultLocale || _zh_CN.default;
        var componentLocale = globalLocale[name || WrappedComponent.name];
        var localeCode = globalLocale.locale;
        var forwardedRef = props.forwardedRef,
            rest = (0, _objectWithoutProperties2.default)(props, ["forwardedRef"]);
        return _react.default.createElement(WrappedComponent, (0, _extends2.default)({}, rest, {
          ref: forwardedRef,
          locale: componentLocale,
          localeCode: localeCode
        }));
      });
    };

    var forwardRef = function forwardRef(props, ref) {
      return _react.default.createElement(LocaleReceiver, (0, _extends2.default)({}, props, {
        forwardedRef: ref
      }));
    };

    var LocaleReceiverWithRef = _react.default.forwardRef(forwardRef);

    (0, _hoistNonReactStatics.default)(LocaleReceiverWithRef, WrappedComponent);
    return LocaleReceiverWithRef;
  };
};

var _default = LocaleReceiverWrapper;
exports.default = _default;
