// zarm/lib/input/Input.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _InputNumber = _interopRequireDefault(require("./InputNumber"));

var _InputBase = _interopRequireDefault(require("./InputBase"));

var _InputTextarea = _interopRequireDefault(require("./InputTextarea"));

var Input =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(Input, _PureComponent);

  function Input() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Input);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Input)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.input = void 0;
    return _this;
  }

  (0, _createClass2.default)(Input, [{
    key: "focus",
    value: function focus() {
      if (this.input) {
        this.input.focus();
      }
    }
  }, {
    key: "blur",
    value: function blur() {
      if (this.input) {
        this.input.blur();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          type = _this$props.type,
          rest = (0, _objectWithoutProperties2.default)(_this$props, ["type"]);

      if (type === 'text' && 'rows' in this.props) {
        return _react.default.createElement(_InputTextarea.default, (0, _extends2.default)({
          ref: function ref(ele) {
            _this2.input = ele;
          }
        }, rest));
      }

      switch (type) {
        case 'idcard':
        case 'price':
        case 'number':
          return _react.default.createElement(_InputNumber.default, (0, _extends2.default)({
            ref: function ref(ele) {
              _this2.input = ele;
            }
          }, this.props));

        case 'text':
        case 'search':
        case 'password':
        default:
          return _react.default.createElement(_InputBase.default, (0, _extends2.default)({
            ref: function ref(ele) {
              _this2.input = ele;
            }
          }, this.props));
      }
    }
  }]);
  return Input;
}(_react.PureComponent);

exports.default = Input;
Input.defaultProps = {
  type: 'text'
};
