// zarm/es/locale-receiver/LocaleReceiver.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import hoistNonReactStatic from 'hoist-non-react-statics';
import { LocaleContext } from '../locale-provider/LocaleProvider';
import defaultLocaleData from '../locale-provider/locale/zh_CN';

var LocaleReceiverWrapper = function LocaleReceiverWrapper(name, defaultLocale) {
  return function InnerWrapper(WrappedComponent) {
    var LocaleReceiver = function LocaleReceiver(props) {
      return React.createElement(LocaleContext.Consumer, null, function (locale) {
        var globalLocale = locale.locale ? locale : defaultLocale || defaultLocaleData;
        var componentLocale = globalLocale[name || WrappedComponent.name];
        var localeCode = globalLocale.locale;

        var forwardedRef = props.forwardedRef,
            rest = _objectWithoutProperties(props, ["forwardedRef"]);

        return React.createElement(WrappedComponent, _extends({}, rest, {
          ref: forwardedRef,
          locale: componentLocale,
          localeCode: localeCode
        }));
      });
    };

    var forwardRef = function forwardRef(props, ref) {
      return React.createElement(LocaleReceiver, _extends({}, props, {
        forwardedRef: ref
      }));
    };

    var LocaleReceiverWithRef = React.forwardRef(forwardRef);
    hoistNonReactStatic(LocaleReceiverWithRef, WrappedComponent);
    return LocaleReceiverWithRef;
  };
};

export default LocaleReceiverWrapper;
