// zarm/es/tabs/TabPanel.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React, { PureComponent } from 'react';
import classnames from 'classnames';

var TabPanel =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(TabPanel, _PureComponent);

  function TabPanel(props) {
    var _this;

    _classCallCheck(this, TabPanel);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(TabPanel).call(this, props));
    _this.state = {
      selected: props.selected
    };
    return _this;
  }

  _createClass(TabPanel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          children = _this$props.children;
      var selected = this.state.selected;
      var cls = classnames("".concat(prefixCls, "__panel"), className, _defineProperty({}, "".concat(prefixCls, "__panel--active"), selected));
      return React.createElement("div", {
        className: cls,
        role: "tabpanel"
      }, children);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps) {
      if ('selected' in nextProps) {
        return {
          selected: nextProps.selected
        };
      }

      return null;
    }
  }]);

  return TabPanel;
}(PureComponent);

TabPanel.defaultProps = {
  prefixCls: 'za-tabs'
};
export { TabPanel as default };
