// zarm/es/confirm/Confirm.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React, { PureComponent } from 'react';
import classnames from 'classnames';
import Modal from '../modal';
import confirmLocale from './locale/zh_CN';

var Confirm =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Confirm, _PureComponent);

  function Confirm() {
    _classCallCheck(this, Confirm);

    return _possibleConstructorReturn(this, _getPrototypeOf(Confirm).apply(this, arguments));
  }

  _createClass(Confirm, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          content = _this$props.content,
          okText = _this$props.okText,
          cancelText = _this$props.cancelText,
          onOk = _this$props.onOk,
          onCancel = _this$props.onCancel,
          locale = _this$props.locale,
          others = _objectWithoutProperties(_this$props, ["prefixCls", "className", "content", "okText", "cancelText", "onOk", "onCancel", "locale"]);

      var cls = classnames(prefixCls, className);
      return React.createElement(Modal, _extends({
        className: cls
      }, others, {
        footer: React.createElement(React.Fragment, null, React.createElement("button", {
          type: "button",
          className: "".concat(prefixCls, "__button"),
          onClick: onCancel
        }, cancelText || locale.cancelText), React.createElement("button", {
          type: "button",
          className: "".concat(prefixCls, "__button ").concat(prefixCls, "__button--ok"),
          onClick: onOk
        }, okText || locale.okText))
      }), content);
    }
  }]);

  return Confirm;
}(PureComponent);

Confirm.defaultProps = {
  prefixCls: 'za-confirm',
  animationType: 'zoom',
  locale: confirmLocale
};
export { Confirm as default };
