// zarm/lib/tabs/TabPanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var TabPanel =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(TabPanel, _PureComponent);

  function TabPanel(props) {
    var _this;

    (0, _classCallCheck2.default)(this, TabPanel);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TabPanel).call(this, props));
    _this.state = {
      selected: props.selected
    };
    return _this;
  }

  (0, _createClass2.default)(TabPanel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          prefixCls = _this$props.prefixCls,
          className = _this$props.className,
          children = _this$props.children;
      var selected = this.state.selected;
      var cls = (0, _classnames2.default)("".concat(prefixCls, "__panel"), className, (0, _defineProperty2.default)({}, "".concat(prefixCls, "__panel--active"), selected));
      return _react.default.createElement("div", {
        className: cls,
        role: "tabpanel"
      }, children);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps) {
      if ('selected' in nextProps) {
        return {
          selected: nextProps.selected
        };
      }

      return null;
    }
  }]);
  return TabPanel;
}(_react.PureComponent);

exports.default = TabPanel;
TabPanel.defaultProps = {
  prefixCls: 'za-tabs'
};
